### ISSUE TITLE
Training set NDArrayIter empty after first epoch

### ISSUE THREAD

**Issue by nickeleres** [apache/incubator-mxnet#8370]
The code contained within the loop `for i, batch in enumerate(train_data):` is only run once, which I assume is because for some reason `train_data` is empty/undefined after the first full pass thru the data (first epoch).  

```
epochs = 10
for e in range(epochs):
    print 'Epoch:', e, 'train_data', enumerate(train_data)
    for i, batch in enumerate(train_data):
        print 'epoch', e, 'batch_number', i, 'train_data', train_data
        data = batch.data[0]
        label = batch.label[0]
        with autograd.record(): # Start recording the derivatives
            output = net(data) # the forward iteration
            loss = softmax_cross_entropy(output, label)
            epoch_loss = loss.asnumpy().sum()/batch_size
            loss.backward()
        trainer.step(data.shape[0])
```

I have to redefine `train_data` (below) at the end of the `for e in range(epochs)` loop in order for every epoch to run:

`train_data = mx.io.NDArrayIter(train_data_mx, train_label_mx, batch_size, shuffle=True)`

Am I missing a flag somewhere?  Why would this be happening?

**Comment by reminisce** [comment]
Try replacing that with `train_data.reset()`.

**Comment by nickeleres** [comment]
Doesnt appear to work

**Comment by reminisce** [comment]
Could you post the code after you replacing with `train_data.reset()`?

**Comment by nickeleres** [comment]
`loss_seq_train = []
loss_seq_test = []
acc_seq_train = []
acc_seq_test = []

epochs = 30
for e in range(epochs):
    print 'Training epoch', e+1
    for i, batch in enumerate(train_data):
        data = batch.data[0]
        label = batch.label[0]
        with autograd.record(): # Start recording the derivatives
            output = net(data) # the forward iteration
            loss = softmax_cross_entropy(output, label)
            epoch_loss = loss.asnumpy().sum()/batch_size
            print 'epoch_loss', epoch_loss
            loss.backward()
        trainer.step(data.shape[0])
            
    # train_data = mx.io.NDArrayIter(train_data_mx, train_label_mx, batch_size, shuffle=True)
    train_data.reset()
    # test_data = mx.io.NDArrayIter(test_data_mx, test_label_mx, batch_size, shuffle=True)
     test_data.reset()
    
    # Accuracy metrics
    test_accuracy, test_loss = evaluate_accuracy(test_data, net)
    train_accuracy, train_loss = evaluate_accuracy(train_data, net)

    # save them for later
    loss_seq_train.append(train_loss)
    loss_seq_test.append(test_loss)
    acc_seq_train.append(train_accuracy)
    acc_seq_test.append(test_accuracy)

    if True:
        print("Completed epoch %s. Train Loss: %s, Test Loss %s, Train_acc %s, Test_acc %s" %
              (e+1, train_loss, test_loss, train_accuracy, test_accuracy))`

**Comment by reminisce** [comment]
What does `evaluate_accuracy` do? Does it change the state of `train_data`? Could you post the code as well?

**Comment by nickeleres** [comment]
```
def evaluate_accuracy(data_iterator, net):
    numerator = 0.
    denominator = 0.
    loss_avg = 0.
    for i, batch in enumerate(data_iterator):
        data = batch.data[0]
        label = batch.label[0]
        output = net(data)
        loss = softmax_cross_entropy(output, label)
#         loss = softmax_cross_entropy(output, label_one_hot)
        predictions = nd.argmax(output, axis=1)
        predictions = nd.one_hot(predictions, num_outputs)
#         print 'predictions', predictions, 'label', label
        numerator += nd.sum(predictions == label)
        denominator += data.shape[0]
        loss_avg = loss_avg*i/(i+1) + nd.mean(loss).asscalar()/(i+1)
    return (numerator / denominator).asscalar(), loss_avg
```

**Comment by reminisce** [comment]
Looks like you changed the state of `train_data` in `evaluate_accuracy`. You need to reset it again after that function call. Same for other data iters.

**Comment by nickeleres** [comment]
Huh?...so the state of `train_data` is globally mutable from within the scope of another function?

Also, this code is nearly line-for-line from The Straight Dope documentation...

http://gluon.mxnet.io/chapter02_supervised-learning/regularization-gluon.html

**Comment by reminisce** [comment]
Passing an object to a function in python is like passing an object's pointer to a function in C++. So here `train_iter` and `data_iterator` point to the same memory space.

**Comment by reminisce** [comment]
I guess the difference is in the tutorial, it's a `DataLoader` type, which is not an iterator, while you are using a iterator.

**Comment by nickeleres** [comment]
Ok thank you for clarifying.  Apologies for the false alarm.
